nflow-engine/src/main/java/io/nflow/engine/service/ArchiveService.java

```diff
@@ -40,34 +50,162 @@ public ArchiveService(ArchiveDao archiveDao) {
+              source, archiveDao::archiveWorkflows);
+    }
+    if (configuration.deleteWorkflowsOlderThan != null) {
+      Supplier<List<Long>> source = getIdQuery(MAIN, configuration, configuration.deleteWorkflowsOlderThan);
+      res.deletedWorkflows = doAction("Deleting workflows", format("Deleting passive workflows older than %s, in batches of %s.", configuration.deleteWorkflowsOlderThan, configuration.batchSize),
+              source, idList -> archiveDao.deleteWorkflows(MAIN, idList));
+    }
+    if (configuration.deleteStatesOlderThan != null) {
+      // TODO
+    }
+    return res;
+  }
+
+  private Supplier<List<Long>> getIdQuery(TablePrefix table, ArchiveConfiguration configuration, Duration duration) {
+    DateTime olderThan = now().minus(duration);
+    return configuration.operateOnHierarchies ?
+              () -> archiveDao.listOldWorkflowTrees(table, olderThan, configuration.batchSize) :
+              () -> archiveDao.listOldWorkflows(table, olderThan, configuration.batchSize);
+  }
+
+  private int doAction(String type, String description, Supplier<List<Long>> getActionables, Function<List<Long>, Integer> doAction) {
+    log.info(description);
     StopWatch stopWatch = new StopWatch();
     stopWatch.start();
-    List<Long> workflowIds;
     PeriodicLogger periodicLogger = new PeriodicLogger(log, 60);
-    int archivedWorkflowsTotal = 0;
+    int totalWorkflows = 0;
     do {
-      workflowIds = archiveDao.listArchivableWorkflows(olderThan, batchSize);
+      List<Long> workflowIds = getActionables.get();
       if (workflowIds.isEmpty()) {
         break;
       }
-      int archivedWorkflows = archiveDao.archiveWorkflows(workflowIds);
-      archivedWorkflowsTotal += archivedWorkflows;
+      workflowIds.sort(null);
+      int workflows = doAction.apply(workflowIds);
+      totalWorkflows += workflows;
 
       double timeDiff = max(stopWatch.getTime() / 1000.0, 0.000001);
-      log.debug("Archived {} workflows. {} workflows / second. Workflow ids: {}. ", archivedWorkflows, archivedWorkflowsTotal
-          / timeDiff, workflowIds);
-      periodicLogger.info("Archived {} workflows. Archiving about {} workflows / second.", archivedWorkflows,
-          archivedWorkflowsTotal / timeDiff);
+      String status = format("%s. %s workflows, %.1f workflows / second.", type, workflows, totalWorkflows / timeDiff);
+      if (log.isDebugEnabled()) {
+        log.debug(status + " Workflow ids: {}.", workflowIds);
+      }
+      periodicLogger.info(status);
     } while (true);
 
-    log.info("Archiving finished. Archived {} workflows in {} seconds.", archivedWorkflowsTotal, stopWatch.getTime() / 1000);
-    return archivedWorkflowsTotal;
+    log.info("{} finished. Operated on {} workflows in {} seconds.", type, totalWorkflows, stopWatch.getTime() / 1000);
+    return totalWorkflows;
+  }
+
+  public static class ArchiveConfiguration {
+    private final Duration deleteArchivedWorkflowsOlderThan;
```

Comment:
these fields could be public? at least package protected to avoid synthetic accessors